src/types/utils.ts

```diff
@@ -257,3 +245,152 @@ export function typeNameToTypeNode(astT: TypeName): TypeNode {
 
     throw new Error(`NYI converting AST Type ${astT.print()} to SType`);
 }
+
+/**
+ * Computes a `TypeNode` equivalent of given `astT`,
+ * specialized for location `loc` (if applicable).
+ */
+export function typeNameToSpecializedTypeNode(astT: TypeName, loc: DataLocation): TypeNode {
+    return specializeType(typeNameToTypeNode(astT), loc);
+}
+
+/**
+ * Given a `VariableDeclaration` node `decl` compute the `TypeNode` that corresponds to the variable.
+ * This takes into account the storage location of the `decl`.
+ */
+export function variableDeclarationToTypeNode(decl: VariableDeclaration): TypeNode {
+    assert(decl.vType !== undefined, "Expected {0} to have type", decl);
+
+    return typeNameToSpecializedTypeNode(decl.vType, decl.storageLocation);
+}
+
+export function enumToIntType(decl: EnumDefinition): IntType {
+    const length = decl.children.length;
+
+    let size: number | undefined;
+
+    for (let n = 8; n <= 32; n += 8) {
+        if (length < 2 ** n) {
+            size = n;
+
+            break;
+        }
+    }
+
+    assert(
+        size !== undefined,
+        "Unable to detect enum type size - member count exceeds 2 ** 32",
+        decl
+    );
+
+    return new IntType(size, false);
+}
+
+/**
+ * Converts `type` to value type. Following rules are applied:
+ * 1. Contract definitions turned to address.
+ * 2. Enum definitions turned to uint of minimal fitting size.
+ * 3. Struct definitions turned to tuples.
+ *   3.1. Mappings are completely ommited.
+ *   3.2. Arrays of root structs are ommited, while preserved for nested structs.
+ * 4. Data locations are skipped.
+ */
+function toValueType(type: TypeNode, depth = 0): TypeNode {
+    if (type instanceof PointerType) {
+        return toValueType(type.to, depth);
+    }
```

Comment:
@cd1m0 I wonder if this is valid. Maybe we should skip `specializeType()` in downstream instead.

Reply:
I would prefer to generalize types here, and keep types "specialized" everywhere else. Both maru-ir and scribble depend on `variableDeclarationToTypeNode` to return a specialized type, and they specialize the return of `typeNameToTypeNode` whenever they use it. It seems that we only generalize types when we use them for something web3 related